### ISSUE TITLE
Unable to start 2 tidb servers on same node when leaving 'socket' config empty(auto-generated)

### ISSUE THREAD

**Issue by bb7133** [pingcap/tidb#29307]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Trying to start 2 tidb servers with different port and 'status port', leaving 'socket' unspecified, for example:

```
bash> tidb-server -P 4000 -status 10080
bash> tidb-server -P 4001 -status 10081
```

### 2. What did you expect to see? (Required)
TiDB Server starts successfully.

### 3. What did you see instead (Required)
The following error for the second TiDB server:
```
[2021/11/01 08:57:20.790 +00:00] [FATAL] [main.go:670] ["failed to create the server"] [error="unix socket /tmp/tidb.sock exists and is functional, not removing it"] 
[stack="main.createServer\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/tidb-server/main.go:670\nmain.main\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/tidb-server/main.go:199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"] 
[stack="main.createServer\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/tidb-server/main.go:670\nmain.main\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/tidb-server/main.go:199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225"]
```

### 4. What is your TiDB version? (Required)
latest master(ad8534185781e3fa19d655885c0ae91bf777080c).

**Comment by bb7133** [comment]
It can be treated as either a bug or 'breaking change', say if we leave the 'port' of TiDB default, the second TiDB server cannot be started either.

However, since we may have a lot of users deployed multiple TiDBs on the same node already, upgrading their clusters to the next release may be failed due to this issue.

I think an easy way to solve it, is adding the port number to the suffix of the default socket file like('`/tmp/tidb-4000.sock`' and '`tmp/tidb-4001.sock`'), @morgo @dveeden do you think it's reasonable?

**Comment by dveeden** [comment]
This is basically what https://github.com/pingcap/tiup/pull/1593 does for TiUP.

I think that it is up to the tool (or person) managing multiple instances to make sure the TCP port, data directory and UNIX socket path are unique. However I do see how this change impacts upgrades.

Having a more dynamic socket file path might make it more difficult for tools to find it and embedding the TCP port in there means that changing TCP settings also affect non-TCP connections, which isn't great.

So the possible solutions are:

1. Not doing anything. Result: compatibility/upgrade issues for running multiple instances on the same host outside of TiUP.
2. Adding the TCP port number in the socket path. Result: less predictable socket file location (affects documentation etc)
3. Making the error for a socket file that is in use non fatal
4. Using `/tmp/tidb-<nr>.sock` if opening the socket file fails (and nr++ until success)

**Comment by morgo** [comment]
My preference is also (1). We only officially support TiUP and operator - so these need to correctly handle the change, but for manual install users will need to adapt. So it's important this is pointed out in the release notes, and we never cherry pick.

**Comment by bb7133** [comment]
Option 1 requires the update for TiUP, it do introduce errors for users with the old version of TiUP, this is the common case we need to consider IMO.

I still prefer option 2, and the socket file is 'not so hard to predict' since the port is easy to get.

**Comment by morgo** [comment]
OK lets change it to option 2. We don't have much time to decide, since it should be included in TiDB 5.3.
